xds/src/main/java/io/grpc/xds/client/XdsClientImpl.java

```diff
@@ -747,4 +983,134 @@ private void notifyWatcher(ResourceWatcher<T> watcher, T update) {
+      checkNotNull(xdsResourceType, "xdsResourceType");
+      syncContext.throwIfNotInThisSynchronizationContext();
+      Set<String> toParseResourceNames =
+          xdsResourceType.shouldRetrieveResourceKeysForArgs()
+          ? getResourceKeys(xdsResourceType)
+          : null;
+      XdsResourceType.Args args = new XdsResourceType.Args(serverInfo, versionInfo, nonce,
+          bootstrapInfo, securityConfig, toParseResourceNames);
+      handleResourceUpdate(args, resources, xdsResourceType, processingTracker);
+    }
+
+    @Override
+    public void handleStreamClosed(Status status, boolean inRetry,
+                                   ControlPlaneClient cpcThatClosed) {
+      syncContext.throwIfNotInThisSynchronizationContext();
+      if (!inRetry) {
+        cleanUpResourceTimers(cpcThatClosed);
+      }
+
+      Status error = status.isOk() ? Status.UNAVAILABLE.withDescription(CLOSED_BY_SERVER) : status;
+      boolean checkForFallback = !status.isOk() && BootstrapperImpl.isEnabledXdsFallback()
+          && !inRetry;
+
+      List<String> authoritiesForClosedCpc = getActiveAuthorities(cpcThatClosed);
+      for (Map<String, ResourceSubscriber<? extends ResourceUpdate>> subscriberMap :
+          resourceSubscribers.values()) {
+        for (ResourceSubscriber<? extends ResourceUpdate> subscriber : subscriberMap.values()) {
+          if (subscriber.hasResult() || !authoritiesForClosedCpc.contains(subscriber.authority)) {
+            continue;
+          }
+
+          // try to fallback to lower priority control plane client
+          if (checkForFallback && doFallbackForAuthority(
+              cpcThatClosed, serverInfo, subscriber.serverInfos, subscriber.authority)) {
+            authoritiesForClosedCpc.remove(subscriber.authority);
+            if (authoritiesForClosedCpc.isEmpty()) {
+              return; // optimization: no need to continue once all authorities have done fallback
+            }
+            continue;
+          }
+
+          subscriber.onError(error, null);
+        }
+      }
+    }
+
+    @Override
+    public void handleStreamRestarted(ServerInfo serverInfo) {
+      if (isShutDown()) {
+        return;
+      }
+
+      syncContext.throwIfNotInThisSynchronizationContext();
+      ControlPlaneClient controlPlaneClient = serverCpClientMap.get(serverInfo);
+      if (controlPlaneClient == null) {
+        return;
+      }
+
+      for (String authority : getAuthoritiesForServerInfo(serverInfo)) {
+        if (controlPlaneClient.hasSubscribedResources(authority)) {
+          internalHandleStreamReady(serverInfo, controlPlaneClient, authority);
+        }
+      }
+    }
+
+    private void internalHandleStreamReady(
+        ServerInfo serverInfo, ControlPlaneClient controlPlaneClient, String authority) {
+      ControlPlaneClient activeCpClient = getActiveCpc(authority);
+      addCpcToAuthority(authority, controlPlaneClient);
```

Comment:
Why are we adding to the list here? Shouldn't we have done that in `fallBackToCpc()`? It looks like if it isn't registered for the authority yet, then that means it _shouldn't_ be added because a higher priority server is being used for that authority.

Using `getAuthoritiesForServerInfo()` before getting here looks to be a mistake. It should be based on `activeCpClients`, not `bootstrapinfo`.

Reply:
There is a need to add the first one in this method.  You are right that it is cleaner to make that explicit rather than use the general method which could potentially cover up some problem in the fallback logic.

Reply:
> There is a need to add the first one in this method.

Is there? Why would the first be special? I see nothing that would guarantee this is the first ServerInfo in the authority's list, so it seems this could also make things out-of-order.

We're looping through all authorities that this serverInfo can potentially serve, not just the ones it has been requested on. By definition if it isn't in the list already it isn't supposed to be answering for this authority. So it seems clearly wrong to go about adding to the list.

We got here because the authority has subscribers. Is it even possible for an authority to have a subscriber, non-empty serverInfos, and yet not have any cpc in the list? It looks like `watchXdsResource()` would add to the list (via `addCpcToAuthority()`) when the subscriber was created.

I don't see the point of using `getAuthoritiesForServerInfo()` (which loops through all server info registrations) and then figure out if we care about the authority with `hasSubscribers()` and `activeCpClients.get(authority).contains(readyCpc)`. Seems all we need to do is:
```java
for (Map.Entry<String, List<ControlPlaneClient>> me : activeCpClients.entrySet()) {
  if (me.getValue().contains(controlPlaneClient)) {
    internalHandleStreamReady(controlPlaneClient, me.getValue());
  }
}
```

There's not need to check `isShutDown()` as shut down cleans up `activeCpClients`. With this construction, it is impossible for `getActiveCpc()` to return `null`. There's no need to check `activeCpClient.isReady() && !activeCpClients.get(authority).contains(readyCpc)`.

Reply:
You are correct that it doesn't need to be added here.  The cpc would already have been added in `manageControlPlaneClient()` before we did the `adjustResourceSubscription()` that would trigger the stream ready.

Added clearing out `activatedCpClients` in `XdsClient.shutdown()` to be able to remove the shutdown check